Implemement TypeSystemSwiftTypeRef::GetBitSize() (NFC)

adrian-prantl opened the pull request:
In order to implement GetBitSize, this patch also adds an initial
TypeRef-based implementation of
SwiftLanguageRuntimeImpl::DoArchetypeBinding(). To make
SwiftLanguageRuntimeImpl::GetTypeInfo() work with TypeRefs of
ClangImported non-Objective-C types, a new callback is added to
swift::reflection::MemoryReader, which allos LLDB to provide the
swift::reflection::TypeInfo for a Clang type based on debug
information.

<rdar://problem/55412920>

Needs https://github.com/apple/swift/pull/33417

adrian-prantl committed:
Implemement TypeSystemSwiftTypeRef::GetBitSize()  (NFC)

In order to implement GetBitSize, this patch also adds an initial
TypeRef-based implementation of
SwiftLanguageRuntimeImpl::DoArchetypeBinding(). To make
SwiftLanguageRuntimeImpl::GetTypeInfo() work with TypeRefs of
ClangImported non-Objective-C types, a new callback is added to
swift::reflection::MemoryReader, which allos LLDB to provide the
swift::reflection::TypeInfo for a Clang type based on debug
information.

<rdar://problem/55412920>

adrian-prantl commented:
This is *not* yet ready, a few tests are still failing. Just trying to get early feedback for https://github.com/apple/swift/pull/33417

adrian-prantl commented:
@kastiglione This is my work-in-progress for the `GetBitSize()` implementation. The patch is rather larger because it depends on an implementation of `DoArchetypBinding()`, which makes up the bulk the wor.

adrian-prantl commented:
This is good to go now.

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
test with https://github.com/apple/swift/pull/33417
@swift-ci test

adrian-prantl commented:
test with https://github.com/apple/swift/pull/33417
@swift-ci test

kastiglione commented on lldb/source/Target/SwiftLanguageRuntimeImpl.h:
```diff
@@ -114,6 +134,21 @@ class SwiftLanguageRuntimeImpl {
 protected:
+  /// Use the reflection context to build a TypeRef object.
+  ///
+  ///\param module can be used to specify a module to look up DWARF
```
is this correct formatting, or will it cause rendering issues?

adrian-prantl replied:
That's a typo!

kastiglione commented on lldb/source/Target/SwiftLanguageRuntimeImpl.h:
```diff
@@ -114,6 +134,21 @@ class SwiftLanguageRuntimeImpl {
+  /// is necessary to specify the module explicitly is if a type has
+  /// been imported into the scratch context. This is always the
+  /// module of the outermost type. Even for bound generic types,
+  /// we're only interested in the module the BGS came from, the bound
```
what is "BGS"?

adrian-prantl replied:
`BoundGenericStruct`, but I probably should have said `BGT`, but really it should be `bound generic type` :-)

kastiglione commented on lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp:
```diff
@@ -89,12 +89,37 @@ GetClangTypeNode(CompilerType clang_type, swift::Demangle::Demangler &Dem) {
+    n = n->getFirstChild();
+  if (n && n->getKind() == Node::Kind::Type && n->hasChildren())
+    n = n->getFirstChild();
+  return n;
```
should this be "all or nothing"? If the given node is a `Global` but does not contain a `TypeMangling`, this function will return some non-`Type` node. If it's expected that `n` is always going to match the specific kind-traversal, should there be asserts that it's well formed?

kastiglione commented on lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp:
```diff
@@ -89,12 +89,37 @@ GetClangTypeNode(CompilerType clang_type, swift::Demangle::Demangler &Dem) {
+static swift::Demangle::NodePointer
+GetDemangledType(swift::Demangle::Demangler &Dem, StringRef name) {
+  NodePointer n = Dem.demangleSymbol(name);
+  return GetType(Dem, n);
```
should `GetType` be inlined here since it's not called elsewhere and this function is small.

kastiglione commented on lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp:
```diff
@@ -617,13 +627,100 @@ class ASTVerifier : public swift::ASTWalker {
+                  mangledName.str().c_str());
+
+    // Materialize a Clang type from the debug info.
+    assert(swift::Demangle::getManglingPrefixLength(mangledName) == 0);
```
should these and other asserts have `&& "some explanation"`?

kastiglione commented on lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp:
```diff
@@ -1027,22 +1155,117 @@ SwiftLanguageRuntimeImpl::GetPromiseForTypeNameAndFrame(const char *type_name,
+  if (!reflection_ctx)
+    return ts.GetTypeFromMangledTypename(mangled_name);
+
+  swift::Demangle::Demangler Dem;
```
many functions declare `using swift::Demangle;`, should this too? Is there a rule being followed for when to and when not to? Should some of these files have a single top level `using` instead of repeated ones?

kastiglione commented on lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp:
```diff
@@ -1027,22 +1155,117 @@ SwiftLanguageRuntimeImpl::GetPromiseForTypeNameAndFrame(const char *type_name,
 
+static void
+ForEachGenericParameter(swift::Demangle::NodePointer node,
+                        std::function<void(unsigned, unsigned)> callback) {
```
can/should this be `llvm::function_ref`?

kastiglione commented on lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp:
```diff
@@ -1027,22 +1155,117 @@ SwiftLanguageRuntimeImpl::GetPromiseForTypeNameAndFrame(const char *type_name,
+  }
+  default:
+    // Visit the child nodes.
+    for (unsigned i = 0; i < node->getNumChildren(); ++i)
+      ForEachGenericParameter(node->getChild(i), callback);
```
I believe this can use range `for`

```suggestion
    for (auto *child : node)
      ForEachGenericParameter(child, callback);
```

kastiglione commented on lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp:
```diff
@@ -1027,22 +1155,117 @@ SwiftLanguageRuntimeImpl::GetPromiseForTypeNameAndFrame(const char *type_name,
 }
 
+static void
+ForEachGenericParameter(swift::Demangle::NodePointer node,
```
Should the name state that it's DFS?

kastiglione commented on lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp:
```diff
@@ -986,7 +1042,11 @@ bool TypeSystemSwiftTypeRef::Verify(opaque_compiler_type_t type) {
-template <typename T> bool Equivalent(T l, T r) { return l == r; }
+template <typename T> bool Equivalent(T l, T r) {
+  if (l != r)
+    llvm::dbgs() <<  l << " != " << r << "\n";
```
do you mean to commit this?

adrian-prantl replied:
This odd function is used by the assertion that is part of the VALIDATE_AND_RETURN macro, that runs each query through TypeSystemSwiftTypeRef and SwiftASTContext and compares the results to make sure the new implementation doesn't diverge from the old one without us knowing and deciding that the divergence is acceptable. Since it's only run inside of assert() we might as well dump useful debug information in the failure case.
The idea is to delete the SwiftASTContext functions once the transition is complete and then this function in all its ugliness will be deleted, too.

kastiglione commented on lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp:
```diff
@@ -1120,7 +1180,23 @@ template <> bool Equivalent<ConstString>(ConstString l, ConstString r) {
+  // thus assume that a larger number is "better".
+  if (l.hasValue() && r.hasValue() && *l > *r)
+    return true;
+  llvm::dbgs() << l << " != " << r << "\n";
```
same, should this be committed?

kastiglione commented on lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp:
```diff
@@ -1027,22 +1155,117 @@ SwiftLanguageRuntimeImpl::GetPromiseForTypeNameAndFrame(const char *type_name,
+  // function, we don't want to do this earlier, because the
+  // canonicalization in GetCanonicalDemangleTree() must be performed in
+  // the original context as to resolve type aliases correctly.
+  Status error;
```
should the error be inspected? either logged or asserted?

adrian-prantl replied:
Since we are just pushing strings around there is no way that this could actually fail with a typeref as input. That said — getting the scratch context could. Fixed!

kastiglione commented on lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp:
```diff
@@ -1714,49 +1937,92 @@ lldb::addr_t SwiftLanguageRuntime::FixupAddress(lldb::addr_t addr,
+    return nullptr;
+  swift::Demangle::NodePointer node =
+      TypeSystemSwiftTypeRef::GetCanonicalDemangleTree(
+          module ? module : ts->GetModule(), Dem, mangled_name.GetStringRef());
```
```suggestion
          module ?: ts->GetModule(), Dem, mangled_name.GetStringRef());
```

kastiglione commented on lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp:
```diff
@@ -1714,49 +1937,92 @@ lldb::addr_t SwiftLanguageRuntime::FixupAddress(lldb::addr_t addr,
+      ExecutionContext exe_ctx;
+      // FIXME: Should be
+      // frame->CalculateExecutionContext(exe_ctx);
+      // but all the other functions currently get this wrong, too!
+      m_process.GetTarget().CalculateExecutionContext(exe_ctx);
```
am I understanding right that this isn't using `frame`, to match other places calling it wrong?

adrian-prantl replied:
Yes, I made a note in <rdar://problem/67587895> Locking the wrong scratch context in most places in SwiftLanguageRuntimeDynamicTypeResolution.cpp to fix this in a separate commit.

kastiglione reviewed (commented):
This is a first pass of reading and questions, there's a lot here (for me anyway), I'll do a follow up read. Sorry for some minor comments.

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -7814,6 +7814,28 @@ bool SwiftASTContext::IsImportedType(opaque_compiler_type_t type,
+
+std::string SwiftASTContext::ImportName(const clang::NamedDecl *clang_decl) {
+  if (auto clang_importer = GetClangImporter()) {
+    // FIMXE: It's not clear why this special case is necessary.
```
Note to self: I forgot about this. I meant to check this before uploading.

adrian-prantl replied:
It's no longer necessary.

adrian-prantl committed:
Factor out TypeSystemSwift::GetSupportedLanguagesForTypes() (NFC)

adrian-prantl committed:
Implement TypeSystemSwiftTypeRef::IsImportedType() (NFC)

The SwiftOptionSet data formatter still reconstructs types after ths
patch because there is leftover work in ClangImporter to separate out
name importing.

<rdar://problem/67709905>

adrian-prantl commented:
test with https://github.com/apple/swift/pull/33417
@swift-ci test

adrian-prantl committed:
Adapt to upstream changes introducing type lookup errors.

adrian-prantl commented:
test with https://github.com/apple/swift/pull/33417
@swift-ci smoke test

adrian-prantl commented:
test with https://github.com/apple/swift/pull/33417
@swift-ci smoke test

adrian-prantl commented:
test with https://github.com/apple/swift/pull/33417
@swift-ci  test

adrian-prantl commented:
test with https://github.com/apple/swift/pull/33417
@swift-ci test

adrian-prantl committed:
TestSwiftSystemFrameworks: Pick another random datatype from the
module that does not have two different implementations that vary
based off an opaque #define.

adrian-prantl commented:
test with https://github.com/apple/swift/pull/33417
@swift-ci test

adrian-prantl committed:
qualify dyn_cast (NFC)

adrian-prantl commented:
test with https://github.com/apple/swift/pull/33417
@swift-ci test

adrian-prantl committed:
Disable reproducer test on Linux

adrian-prantl commented:
test with https://github.com/apple/swift/pull/33417
@swift-ci test

kastiglione commented:
Did this not get picked into master-next?

adrian-prantl commented:
Yes, at https://github.com/apple/llvm-project/commit/dab2fc65fba8f735be7f315c24c587c20525a412 <https://github.com/apple/llvm-project/commit/dab2fc65fba8f735be7f315c24c587c20525a412>
Sorry I forgot to put a note into the PR.

-- adrian

> On Sep 10, 2020, at 12:41 PM, Dave Lee <notifications@github.com> wrote:
> 
> 
> Did this not get picked into master-next?
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub <https://github.com/apple/llvm-project/pull/1636#issuecomment-690675127>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AARFXBOGDBALRC5T2QBUG33SFETVBANCNFSM4P3Z67ZQ>.
>

kastiglione commented:
Do we prefer to push directly instead of using a PR?

kastiglione commented on lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp:
```diff
@@ -1587,7 +1643,64 @@ TypeSystemSwiftTypeRef::GetPointerType(opaque_compiler_type_t type) {
                                    ExecutionContextScope *exe_scope) {
-  return m_swift_ast_context->GetBitSize(ReconstructType(type), exe_scope);
+  auto impl = [&]() -> llvm::Optional<uint64_t> {
+    // Bug-for-bug compatibility. See comment in SwiftASTContext::GetBitSize().
```
How come this implementation is so much different from `master-next`? https://github.com/apple/llvm-project/commit/dab2fc65fba8f735be7f315c24c587c20525a412#diff-4f052b540e66838ef123c3a839bacf1cR1617-R1629

kastiglione replied:
For example `GetAsClangTypeOrNull` exists in this PR but not in the `master-next` version?

adrian-prantl replied:
Thank you for catching this! I mush have cherry-picked an older version of the commit. Wow.

adrian-prantl replied:
Hmm.. no. git diff origin/swift/master doesn't show a difference. Is it possible that this is the state of one of the earlier commits in the PR and not the final one?

kastiglione replied:
the reason I noticed this is because when I cherry picked #1779, there were changes to `GetAsClangTypeOrNull` that did not exist at all on `swift/master-next`.

kastiglione replied:
I am confused 🤔

kastiglione replied:
It seems I have done something slightly wrong in my cherry-pick in #1779.

adrian-prantl commented:
> Do we prefer to push directly instead of using a PR?

Either way fine — I _usually_ create two PRs so I can merge them at the same time once the testing for the master one has succeeded. This was an exception.
